server/src/main/java/org/elasticsearch/cluster/routing/BatchedRerouteService.java

```diff
@@ -158,7 +165,7 @@ public void onFailure(Exception e) {
 
                 @Override
                 public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                    ActionListener.onResponse(currentListeners, newState);
+                    future.addListener(ActionListener.wrap(() -> ActionListener.onResponse(currentListeners, newState)));
```

Comment:
Actually this doesn't quite work because in the tests we randomly throw an exception from a listener's `onFailure` method, and `ActionListener.wrap` asserts that this doesn't happen. I think the assertion is correct and the test is wrong:

```diff
diff --git a/server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java b/server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java
index 69d4b169850..a53bbf3b322 100644
--- a/server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java
@@ -201,17 +201,16 @@ public class BatchedRerouteServiceTests extends ESTestCase {
         final int iterations = between(1, 100);
         final CountDownLatch countDownLatch = new CountDownLatch(iterations);
         for (int i = 0; i < iterations; i++) {
-            batchedRerouteService.reroute("iteration " + i, randomFrom(EnumSet.allOf(Priority.class)), ActionListener.wrap(r -> {
-                countDownLatch.countDown();
-                if (rarely()) {
-                    throw new ElasticsearchException("failure during notification");
-                }
-            }, e -> {
-                countDownLatch.countDown();
-                if (randomBoolean()) {
-                    throw new ElasticsearchException("failure during failure notification", e);
-                }
-            }));
+            batchedRerouteService.reroute(
+                "iteration " + i,
+                randomFrom(EnumSet.allOf(Priority.class)),
+                ActionListener.runAfter(ActionListener.wrap(r -> {
+                    if (rarely()) {
+                        throw new ElasticsearchException("failure during notification");
+                    }
+                    countDownLatch.countDown();
+                }, e -> {}), countDownLatch::countDown)
+            );
             if (rarely()) {
                 clusterService.getMasterService()
                     .setClusterStatePublisher(
```